falcon/cluster/cluster.py

```diff
@@ -2,622 +2,343 @@
-    # may differ across runs. Solution is to sort the Lance dataset before
-    # constructing the index, but this is less efficient for large datasets.
-    # In practice, this doesn't seem to have a significant impact on performance.
-    train_spectra = dataset.sample(sample_size)
-    train_vectors = np.stack(train_spectra["vector"].to_numpy())
-    dim = train_vectors.shape[1]
-    # Create an ANN index using the inner product (proxy for cosine
-    # distance) for fast NN queries.
-    if n_list <= 0:
-        index = faiss.IndexIDMap(faiss.IndexFlatIP(dim))
-    else:
-        index = faiss.IndexIVFFlat(
-            faiss.IndexFlatIP(dim), dim, n_list, faiss.METRIC_INNER_PRODUCT
-        )
-        index.nprobe = min(math.ceil(index.nlist / 8), n_probe)
-    # Compute cluster centroids.
-    # noinspection PyArgumentList
-    index.train(train_vectors)
-
-    # Add the vectors to the index in batches.
-    batch_start = 0
-    for batch in tqdm(
-        dataset.to_batches(max_chunksize=batch_size),
-        desc="Batches added to index",
-        unit="batch",
-    ):
-        batch_spectra = batch.to_pandas().apply(
-            spectrum.df_row_to_spec, axis=1
-        )
-        batch_vectors = []
-        for spec in batch_spectra:
-            batch_vectors.append(spec.vector)
-            precursor_mzs.append(spec.precursor_mz)
-            rts.append(spec.retention_time)
-        batch_stop = batch_start + len(batch_vectors)
-        vectors.extend(batch_vectors)
-
-        batch_vectors = np.stack(batch_vectors)
-        index.add_with_ids(
-            batch_vectors,
-            np.arange(batch_start, batch_stop, dtype=np.int64),
-        )
-        batch_start = batch_stop
-    precursor_mzs = np.asarray(precursor_mzs)
-    rts = np.asarray(rts)
-    # Query the index to calculate NN distances.
-    _dist_mz_interval(
-        index,
-        vectors,
-        precursor_mzs,
-        rts,
-        batch_size,
-        n_neighbors,
-        precursor_tol_mass,
-        precursor_tol_mode,
-        rt_tol,
-        distances,
-        indices,
-        indptr,
+    # Hierarchical clustering using the precomputed pairwise distance matrix.
+    min_samples = 2
+    logger.debug(
+        "Hierarchical clustering (distance_threshold=%.4f, min_samples=%d)",
+        distance_threshold,
+        min_samples,
     )
-    index.reset()
-    indptr_i += n_spectra
-
-    metadata = (
+    # Sort the metadata by increasing precursor m/z for easy subsetting.
+    data = (
         dataset.to_table(
             columns=[
-                "filename",
-                "identifier",
                 "precursor_mz",
+                "precursor_charge",
                 "retention_time",
-            ],
-            scan_in_order=True,
+                "mz",
+                "intensity",
+            ]
         )
         .to_pandas()
-        .rename(columns={"identifier": "spectrum_id"})
+        .reset_index()
```

Comment:
**todo:** You don't actually need to reset the index (duplicating the information into an "index" column and the new index), but can use the index directly instead. So line 102 would become `idx = data.index`.